File: app/src/main/java/in/smslite/activity/SelectContactActivity.java
package in.smslite.activity;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.support.annotation.Nullable;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.telephony.PhoneNumberUtils;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import in.smslite.R;
import in.smslite.adapter.SelectContactAdapter;
import in.smslite.utils.ContactUtils;


/**
 * Created by rahul1993 on 4/27/2018.
 */

public class SelectContactActivity extends AppCompatActivity implements LoaderManager.LoaderCallbacks<Cursor> {
  private static final String TAG = SelectContactActivity.class.getSimpleName();
  @BindView(R.id.recyclerView_select_list_id)
  RecyclerView recyclerView;
  @BindView(R.id.editText_select_list_id)
  EditText editText;
  @BindView(R.id.select_contact_activity_back_arrow)
  ImageView imageView;

  private Context context;
  List<String> list = new ArrayList<>();
  List<String> phoneNumberList = new ArrayList<>();

  // Defines a variable for the search string
  private String mSearchString;
  // Defines the array to hold values that replace the ?
  LoaderManager.LoaderCallbacks callbacks;
  public static SelectContactAdapter selectContactAdapter;


  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    context = this;
    callbacks = this;
    setContentView(R.layout.activity_select_contact2);
    ButterKnife.bind(this);
    imageView.setOnClickListener(v -> onBackPressed());
    LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
    editText.addTextChangedListener(new TextWatcher() {
      @Override
      public void beforeTextChanged(CharSequence s, int start, int count, int after) {

      }

      @Override
      public void onTextChanged(CharSequence s, int start, int before, int count) {

      }

      @Override
      public void afterTextChanged(Editable s) {
        StringBuilder text = new StringBuilder();
        text.append(s);
        mSearchString = text.toString();
        getSupportLoaderManager().restartLoader(0, null, callbacks);
//        getLoaderManager().initLoader(0,null, callbacks);
        Log.d(TAG, "afterTextChsnged");
      }
    });

    recyclerView.setHasFixedSize(true);
    recyclerView.setLayoutManager(linearLayoutManager);
//    recyclerView.setLayoutManager(new WrapContentLinearLayoutManager(context));
    selectContactAdapter = new SelectContactAdapter(list, phoneNumberList);
    recyclerView.setAdapter(selectContactAdapter);
  }


  @Override
  public Loader<Cursor> onCreateLoader(int id, Bundle args) {
    phoneNumberList.clear();
    list.clear();
    if (!mSearchString.matches("[0-9]*")) {
      Log.d(TAG, mSearchString + " string");

      String[] projection = {
          ContactsContract.Data._ID,
          ContactsContract.Contacts.HAS_PHONE_NUMBER,
          ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY,
          ContactsContract.CommonDataKinds.Phone.NUMBER
      };
      String selection = ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY + " LIKE ? "+
//          + " OR "+ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY + " LIKE ?" +
          " AND " + ContactsContract.Data.MIMETYPE + " LIKE ?";
      String[] selectionCriteria = {"%"+mSearchString+"%",
          ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE};


      CursorLoader cursor = new CursorLoader(this, ContactsContract.Data.CONTENT_URI, projection, selection,
          selectionCriteria, ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY + " ASC" + " LIMIT 200 ");
      Log.d(TAG, "onCreateLoader");
      return cursor;
    } else {
      Uri uri = ContactsContract.CommonDataKinds.Phone.CONTENT_URI;
      String selection = ContactsContract.CommonDataKinds.Phone.NUMBER + " LIKE?" + " OR "
          + ContactsContract.CommonDataKinds.Phone.NUMBER + " LIKE?" + " OR "
          + ContactsContract.CommonDataKinds.Phone.NUMBER + " LIKE?" + " OR "
          +ContactsContract.CommonDataKinds.Phone.NUMBER + " LIKE?";
//      String[] arg = {"_" + mSearchString + "%" + " OR " + "___" + mSearchString + "%" + " OR " + mSearchString + "%"};
      int length = mSearchString.length();
      String newSearchString = null;
      StringBuilder builder = new StringBuilder();
      for(int i = 0; i < length; i++){
        newSearchString = mSearchString.substring(i, i+1);
        builder.append(newSearchString);
        if(builder.length() == 2 || builder.length() == 5 || builder.length() == 8 || builder.length() == 11){
          builder.append("%");
        }
      }
      Log.d(TAG, builder.toString());
      String[] arg = {"+91" + builder.toString() + "%", "+91 "+ builder.toString()+"%", builder.toString() + "%", "0" + builder.toString() + "%"};
      CursorLoader cursor = new CursorLoader(context, uri, null, selection, arg,ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY + " ASC" + " LIMIT 200 ");
      return cursor;
    }
  }


  @Override
  public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
    if (mSearchString.equals("")) {
      SelectContactAdapter.updateList(list, phoneNumberList);
    } else {
      Log.d(TAG, Integer.toString(data.getCount()));
      if (!mSearchString.matches("[0-9]*")) {
        textIsAlphaNumeric(data);
      } else {
        textIsNumber(data);
      }
    }
  }

  private void textIsNumber(Cursor cursor) {
//    String number = null;
    List<String> nameList = new ArrayList<>();
    List<String> numberList = new ArrayList<>();
    HashSet<String> set = new HashSet<>();
    numberList.add(mSearchString);
    nameList.add(mSearchString);
    try {
      cursor.moveToFirst();
      int nameIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME);
      int numberIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER);
      do {
        String formatedNumber = ContactUtils.normalizeNumber(cursor.getString(numberIndex));
//        if (set.add(formatNumber(cursor.getString(numberIndex)))) {
        if (set.add(formatedNumber)) {
          nameList.add(cursor.getString(nameIndex));
          numberList.add(formatedNumber);
        }
      } while (cursor.moveToNext());
    } catch (Exception e) {
      e.printStackTrace();
    } finally {
      cursor.close();
    }
    SelectContactAdapter.updateList(nameList, numberList);
  }


  private void textIsAlphaNumeric(Cursor cursor) {
    List<String> nameList = new ArrayList<>();
    List<String> numberList = new ArrayList<>();
    HashSet<String> set = new HashSet<>();
    cursor.moveToFirst();

    try {
      int nameIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME);
      int numberIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER);
      do {
//          if (cursor.getInt(1) == 1) {
        if (!cursor.getString(numberIndex).matches(".*[a-zA-Z].*")) {
          String formatedNumber = ContactUtils.normalizeNumber(cursor.getString(numberIndex));
          if (set.add(formatedNumber)) {
            nameList.add(cursor.getString(nameIndex));
            numberList.add(formatedNumber);
          }
        }
      } while (cursor.moveToNext());
    } catch (Exception e) {
      e.printStackTrace();
    } finally {
      cursor.close();
    }
    Log.d(TAG, Integer.toString(nameList.size()));
    Log.d(TAG, Integer.toString(numberList.size()));
    SelectContactAdapter.updateList(nameList, numberList);
  }


  @Override
  public void onLoaderReset(Loader<Cursor> loader) {

  }

  private String formatNumber(String number) {
    number = number.replaceAll("-", "");
    number = number.replaceAll(" ", "");
    if (number.charAt(0) == '+') {
      number = number.substring(3);
      return number;
    }
    if (number.charAt(0) == '0') {
      number = number.substring(1);
      return number;
    }
    return number;
  }

  public class WrapContentLinearLayoutManager extends LinearLayoutManager {

    public WrapContentLinearLayoutManager(Context context) {
      super(context);
    }

    @Override
    public void onLayoutChildren(RecyclerView.Recycler recycler, RecyclerView.State state) {
      try {
        super.onLayoutChildren(recycler, state);
      } catch (IndexOutOfBoundsException e) {
        Log.e("probe", "meet a IOOBE in recycler view");
      }
    }
  }
}

File: app/src/main/java/in/smslite/activity/SelectContactActivity.kt
package `in`.smslite.activity

import `in`.smslite.R
import `in`.smslite.adapter.SelectContactAdapter
import `in`.smslite.databinding.ActivitySelectContact2Binding
import `in`.smslite.utils.ContactUtils
import android.content.Context
import android.database.Cursor
import android.os.Bundle
import android.provider.ContactsContract
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Recycler
import java.util.*
import kotlin.collections.ArrayList

/**
 * Created by rahul1993 on 4/27/2018.
 */
class SelectContactActivity : AppCompatActivity(), LoaderManager.LoaderCallbacks<Cursor> {
    private var context: Context? = null
    var list: ArrayList<String?>? = ArrayList()
    var phoneNumberList: ArrayList<String?>? = ArrayList()
    private lateinit var selectContactAdapter: SelectContactAdapter

    // Defines a variable for the search string
    private var mSearchString: String? = null

    // Defines the array to hold values that replace the ?
    var callbacks: LoaderManager.LoaderCallbacks<*>? = null
    private var binding: ActivitySelectContact2Binding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        context = this
        callbacks = this
        binding = ActivitySelectContact2Binding.inflate(layoutInflater)
        setContentView(R.layout.activity_select_contact2)
        binding!!.selectContactActivityBackArrow.setOnClickListener { v: View? -> onBackPressed() }
        val linearLayoutManager = LinearLayoutManager(this)
        binding!!.editTextSelectListId.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                val text = StringBuilder()
                text.append(s)
                mSearchString = text.toString()
                supportLoaderManager.restartLoader<Any>(0, null, callbacks as LoaderManager.LoaderCallbacks<Any>)
                //        getLoaderManager().initLoader(0,null, callbacks);
                Log.d(TAG, "afterTextChsnged")
            }
        })
        binding!!.recyclerViewSelectListId.setHasFixedSize(true)
        binding!!.recyclerViewSelectListId.layoutManager = linearLayoutManager
        //    recyclerView.setLayoutManager(new WrapContentLinearLayoutManager(context));
        selectContactAdapter = SelectContactAdapter(list, phoneNumberList)
        binding!!.recyclerViewSelectListId.adapter = selectContactAdapter
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {
        phoneNumberList?.clear()
        list?.clear()
        return if (!mSearchString!!.matches(Regex("[0-9]*"))) {
            Log.d(TAG, "$mSearchString string")
            val projection = arrayOf(
                    ContactsContract.Data._ID,
                    ContactsContract.Contacts.HAS_PHONE_NUMBER,
                    ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY,
                    ContactsContract.CommonDataKinds.Phone.NUMBER
            )
            val selection = ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY + " LIKE ? " +  //          + " OR "+ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY + " LIKE ?" +
                    " AND " + ContactsContract.Data.MIMETYPE + " LIKE ?"
            val selectionCriteria = arrayOf("%$mSearchString%",
                    ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
            val cursor = CursorLoader(this, ContactsContract.Data.CONTENT_URI, projection, selection,
                    selectionCriteria, ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY + " ASC" + " LIMIT 200 ")
            Log.d(TAG, "onCreateLoader")
            cursor
        } else {
            val uri = ContactsContract.CommonDataKinds.Phone.CONTENT_URI
            val selection = (ContactsContract.CommonDataKinds.Phone.NUMBER + " LIKE?" + " OR "
                    + ContactsContract.CommonDataKinds.Phone.NUMBER + " LIKE?" + " OR "
                    + ContactsContract.CommonDataKinds.Phone.NUMBER + " LIKE?" + " OR "
                    + ContactsContract.CommonDataKinds.Phone.NUMBER + " LIKE?")
            //      String[] arg = {"_" + mSearchString + "%" + " OR " + "___" + mSearchString + "%" + " OR " + mSearchString + "%"};
            val length = mSearchString!!.length
            var newSearchString: String? = null
            val builder = StringBuilder()
            for (i in 0 until length) {
                newSearchString = mSearchString!!.substring(i, i + 1)
                builder.append(newSearchString)
                if (builder.length == 2 || builder.length == 5 || builder.length == 8 || builder.length == 11) {
                    builder.append("%")
                }
            }
            Log.d(TAG, builder.toString())
            val arg = arrayOf("+91$builder%", "+91 $builder%", "$builder%", "0$builder%")
            CursorLoader(context!!, uri, null, selection, arg, ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME_PRIMARY + " ASC" + " LIMIT 200 ")
        }
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        if (mSearchString == "") {
            selectContactAdapter.updateList(list, phoneNumberList)
        } else {
            Log.d(TAG, Integer.toString(data.count))
            if (!mSearchString!!.matches(Regex("[0-9]*"))) {
                textIsAlphaNumeric(data)
            } else {
                textIsNumber(data)
            }
        }
    }

    private fun textIsNumber(cursor: Cursor) {
//    String number = null;
        val nameList: ArrayList<String?> = ArrayList()
        val numberList: ArrayList<String?> = ArrayList()
        val set = HashSet<String>()
        numberList.add(mSearchString)
        nameList.add(mSearchString)
        try {
            cursor.moveToFirst()
            val nameIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME)
            val numberIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER)
            do {
                val formatedNumber: String = ContactUtils.normalizeNumber(cursor.getString(numberIndex))
                //        if (set.add(formatNumber(cursor.getString(numberIndex)))) {
                if (set.add(formatedNumber)) {
                    nameList.add(cursor.getString(nameIndex))
                    numberList.add(formatedNumber)
                }
            } while (cursor.moveToNext())
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            cursor.close()
        }
        selectContactAdapter.updateList(nameList, numberList)
    }

    private fun textIsAlphaNumeric(cursor: Cursor) {
        val nameList: ArrayList<String?>? = ArrayList()
        val numberList: ArrayList<String?>? = ArrayList()
        val set = HashSet<String>()
        cursor.moveToFirst()
        try {
            val nameIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME)
            val numberIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER)
            do {
//          if (cursor.getInt(1) == 1) {
                if (!cursor.getString(numberIndex).matches(Regex(".*[a-zA-Z].*"))) {
                    val formatedNumber: String = ContactUtils.normalizeNumber(cursor.getString(numberIndex))
                    if (set.add(formatedNumber)) {
                        nameList?.add(cursor.getString(nameIndex))
                        numberList?.add(formatedNumber)
                    }
                }
            } while (cursor.moveToNext())
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            cursor.close()
        }
//        Log.d(TAG, Integer.toString(nameList?.size))
//        Log.d(TAG, Integer.toString(numberList?.size))
        selectContactAdapter.updateList(nameList, numberList)
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {}
    private fun formatNumber(number: String): String {
        var number = number
        number = number.replace("-".toRegex(), "")
        number = number.replace(" ".toRegex(), "")
        if (number[0] == '+') {
            number = number.substring(3)
            return number
        }
        if (number[0] == '0') {
            number = number.substring(1)
            return number
        }
        return number
    }

    inner class WrapContentLinearLayoutManager(context: Context?) : LinearLayoutManager(context) {
        override fun onLayoutChildren(recycler: Recycler, state: RecyclerView.State) {
            try {
                super.onLayoutChildren(recycler, state)
            } catch (e: IndexOutOfBoundsException) {
                Log.e("probe", "meet a IOOBE in recycler view")
            }
        }
    }

    companion object {
        private val TAG = SelectContactActivity::class.java.simpleName

        @JvmField
        var selectContactAdapter: SelectContactAdapter? = null
    }
}